internal/services/tesla_service.go

```diff
@@ -82,3 +86,101 @@ type TeslaVehicle struct {
 	VehicleID int    `json:"vehicle_id"`
 	VIN       string `json:"vin"`
 }
+
+type GetVehiclesResponse struct {
+	Response *[]TeslaVehicle `json:"response"`
+}
+
+type TeslaError struct {
+	Error            string
+	ErrorDescription string `json:"error_description"`
+	ReferenceID      string
+}
+
+var teslaScopes = []string{"user_data", "vehicle_device_data", "vehicle_cmds", "vehicle_charging_cmds", "energy_device_data", "energy_device_data", "energy_cmds"}
+
+type TeslaAuthCodeResponse struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	IDToken      string `json:"id_token"`
+	ExpiresIn    int    `json:"expires_in"`
+	State        string `json:"state"`
+	TokenType    string `json:"token_type"`
+}
+
+func (t *teslaService) CompleteTeslaAuthCodeExchange(authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error) {
+	u, _ := url.Parse("https://auth.tesla.com/oauth2/v3/token")
+	f := url.Values{}
+	f.Set("grant_type", "authorization_code")
+	f.Set("client_id", t.Settings.TeslaAuthorization.ClientID)
+	f.Set("client_secret", t.Settings.TeslaAuthorization.ClientSecret)
+	f.Set("code", authCode)
+	f.Set("redirect_uri", redirectURI)
+	f.Set("scope", "openid offline_access "+strings.Join(teslaScopes, " "))
+	f.Set("audience", fmt.Sprintf("https://fleet-api.prd.%s.vn.cloud.tesla.com", region))
+
+	req, err := http.NewRequest("POST", u.String(), strings.NewReader(f.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		errBody := new(TeslaError)
+		if err := json.NewDecoder(res.Body).Decode(errBody); err != nil {
+			return nil, err
```

Comment:
If we hit this state, the caller should know what happened. As it is this will look like a random decoded error.
```suggestion
			return nil, fmt.Errorf("error occurred completing authorization, body could not be decoded: %w", err)
```

Reply:
You have a point, however this most likely will be an error from a 3rd party and `body could not be decoded` doesn't quite give a clue as to what happened which could be a temporary issue or a more permanent one, plus there is little to nothing the caller can do about this particular error. 

Maybe a better way to handle this is 
- Log the error on the server using a local logger
- change the error message for client to `error occurred completing authorization, invalid response received from during authorization process`

Good observation